File: src/amadeusClient.js
import Amadeus from 'amadeus';
import dotenv from 'dotenv';
import { FlightAPIError, errorCodes } from './errorHandler.js';

dotenv.config();

class AmadeusFlightService {
  constructor() {
    if (!process.env.AMADEUS_API_KEY || !process.env.AMADEUS_API_SECRET) {
      throw new Error('Amadeus API credentials are required. Please set AMADEUS_API_KEY and AMADEUS_API_SECRET in .env file');
    }

    this.amadeus = new Amadeus({
      clientId: process.env.AMADEUS_API_KEY,
      clientSecret: process.env.AMADEUS_API_SECRET,
      hostname: 'test' // Use 'test' for testing environment, 'production' for live
    });
  }

  async searchFlights(searchParams) {
    const { origin, destination, departure_date, return_date, adults, travel_class } = searchParams;

    try {
      const requestParams = {
        originLocationCode: origin,
        destinationLocationCode: destination,
        departureDate: departure_date,
        adults: adults || 1,
        currencyCode: 'USD',
        max: 10 // Limit results to 10 flights
      };

      // Add return date for round-trip flights
      if (return_date) {
        requestParams.returnDate = return_date;
      }

      // Map travel class to Amadeus cabin class
      if (travel_class) {
        const cabinClassMap = {
          'economy': 'ECONOMY',
          'premium_economy': 'PREMIUM_ECONOMY',
          'business': 'BUSINESS',
          'first': 'FIRST'
        };
        requestParams.travelClass = cabinClassMap[travel_class] || 'ECONOMY';
      }

      console.error('Searching flights with Amadeus API:', requestParams);
      
      const response = await this.amadeus.shopping.flightOffersSearch.get(requestParams);
      
      if (!response.data || response.data.length === 0) {
        throw new FlightAPIError(
          'No flights found for the specified criteria',
          errorCodes.FLIGHT_NOT_FOUND,
          { searchParams }
        );
      }

      return this.formatFlightOffers(response.data);

    } catch (error) {
      console.error('Amadeus API Error:', error);

      if (error instanceof FlightAPIError) {
        throw error;
      }

      // Handle Amadeus-specific errors
      if (error.response && error.response.data) {
        const amadeusError = error.response.data;
        
        if (amadeusError.errors && amadeusError.errors.length > 0) {
          const firstError = amadeusError.errors[0];
          
          if (firstError.code === 38189) {
            throw new FlightAPIError(
              'Invalid airport code provided',
              errorCodes.INVALID_AIRPORT_CODE,
              { detail: firstError.detail }
            );
          }
          
          if (firstError.code === 4926) {
            throw new FlightAPIError(
              'Invalid date format or date in the past',
              errorCodes.INVALID_DATE_FORMAT,
              { detail: firstError.detail }
            );
          }

          throw new FlightAPIError(
            firstError.detail || 'Amadeus API error',
            errorCodes.API_UNAVAILABLE,
            { code: firstError.code, detail: firstError.detail }
          );
        }
      }

      // Handle network errors
      if (error.code === 'ENOTFOUND' || error.code === 'ECONNREFUSED') {
        throw new FlightAPIError(
          'Unable to connect to flight data service',
          errorCodes.NETWORK_ERROR,
          { originalError: error.message }
        );
      }

      // Handle rate limiting
      if (error.response && error.response.status === 429) {
        throw new FlightAPIError(
          'API rate limit exceeded. Please try again later',
          errorCodes.API_RATE_LIMIT
        );
      }

      throw new FlightAPIError(
        'Failed to search flights',
        errorCodes.API_UNAVAILABLE,
        { originalError: error.message }
      );
    }
  }

  formatFlightOffers(flightOffers) {
    return flightOffers.map(offer => {
      const itinerary = offer.itineraries[0]; // First itinerary (outbound)
      const segment = itinerary.segments[0]; // First segment
      const lastSegment = itinerary.segments[itinerary.segments.length - 1];
      
      const price = offer.price;
      const validatingAirline = offer.validatingAirlineCodes[0];
      
      // Calculate total duration
      const totalDuration = this.parseDuration(itinerary.duration);
      
      // Count stops
      const stops = itinerary.segments.length - 1;
      
      return {
        airline: this.getAirlineName(validatingAirline),
        airline_code: validatingAirline,
        flight_number: `${segment.carrierCode}${segment.number}`,
        origin: segment.departure.iataCode,
        destination: lastSegment.arrival.iataCode,
        departure_date: segment.departure.at.split('T')[0],
        departure_time: this.formatTime(segment.departure.at),
        arrival_time: this.formatTime(lastSegment.arrival.at),
        duration: totalDuration,
        price: {
          amount: parseFloat(price.total),
          currency: price.currency,
          per_person: parseFloat(price.base),
        },
        stops: stops,
        travel_class: this.mapCabinClass(segment.cabin),
        aircraft: segment.aircraft?.code || 'Unknown',
        booking_class: segment.bookingClass,
        segments: itinerary.segments.length,
        offer_id: offer.id,
      };
    });
  }

  parseDuration(duration) {
    // Duration format: PT4H25M
    const match = duration.match(/PT(?:(\d+)H)?(?:(\d+)M)?/);
    if (!match) return duration;
    
    const hours = parseInt(match[1] || 0);
    const minutes = parseInt(match[2] || 0);
    
    return `${hours}h ${minutes}m`;
  }

  formatTime(dateTimeString) {
    const date = new Date(dateTimeString);
    return date.toISOString().substr(11, 5); // Extract HH:MM
  }

  getAirlineName(airlineCode) {
    const airlineNames = {
      'AA': 'American Airlines',
      'DL': 'Delta Air Lines',
      'UA': 'United Airlines',
      'WN': 'Southwest Airlines',
      'B6': 'JetBlue Airways',
      'AS': 'Alaska Airlines',
      'NK': 'Spirit Airlines',
      'F9': 'Frontier Airlines',
      'G4': 'Allegiant Air',
      'SY': 'Sun Country Airlines',
      'LH': 'Lufthansa',
      'BA': 'British Airways',
      'AF': 'Air France',
      'KL': 'KLM',
      'VS': 'Virgin Atlantic',
      'EK': 'Emirates',
      'QR': 'Qatar Airways',
      'TK': 'Turkish Airlines',
      'SQ': 'Singapore Airlines',
      'CX': 'Cathay Pacific',
      'NH': 'ANA',
      'JL': 'Japan Airlines',
    };
    
    return airlineNames[airlineCode] || airlineCode;
  }

  mapCabinClass(cabinClass) {
    const classMap = {
      'ECONOMY': 'economy',
      'PREMIUM_ECONOMY': 'premium_economy',
      'BUSINESS': 'business',
      'FIRST': 'first'
    };
    
    return classMap[cabinClass] || 'economy';
  }

  async getAirportInfo(airportCode) {
    try {
      console.error(`Getting airport info for: ${airportCode}`);
      
      // Use the locations endpoint instead of airports endpoint
      const response = await this.amadeus.referenceData.locations.get({
        keyword: airportCode,
        subType: 'AIRPORT',
        'page[limit]': 1
      });

      if (!response.data || response.data.length === 0) {
        throw new FlightAPIError(
          `Airport information not found for code: ${airportCode}`,
          errorCodes.INVALID_AIRPORT_CODE,
          { airport_code: airportCode }
        );
      }

      const airport = response.data[0];
      
      return {
        code: airport.iataCode,
        name: airport.name,
        city: airport.address?.cityName || 'Unknown',
        country: airport.address?.countryName || 'Unknown',
        timezone: airport.timeZoneOffset || 'Unknown',
        coordinates: {
          lat: airport.geoCode?.latitude || 0,
          lon: airport.geoCode?.longitude || 0
        }
      };

    } catch (error) {
      console.error('Amadeus Airport Info Error:', error);

      if (error instanceof FlightAPIError) {
        throw error;
      }

      throw new FlightAPIError(
        'Failed to get airport information',
        errorCodes.API_UNAVAILABLE,
        { originalError: error.message }
      );
    }
  }
}

export default AmadeusFlightService;

File: flight_mcp_server/amadeus_client.py
"""Amadeus API client for flight queries."""

import os
import logging
from typing import Dict, List, Optional, Any
from amadeus import Client, ResponseError
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

logger = logging.getLogger(__name__)


class AmadeusFlightService:
    """Service for interacting with Amadeus API."""
    
    def __init__(self):
        """Initialize Amadeus client."""
        self.api_key = os.getenv("AMADEUS_API_KEY")
        self.api_secret = os.getenv("AMADEUS_API_SECRET")
        
        if not self.api_key or not self.api_secret:
            raise ValueError("Amadeus API credentials not found in environment variables")
        
        self.amadeus = Client(
            client_id=self.api_key,
            client_secret=self.api_secret
        )
        logger.info("Amadeus client initialized successfully")
    
    async def search_flights(self, search_params: Dict[str, Any]) -> List[Dict[str, Any]]:
        """Search for flights using Amadeus API."""
        try:
            # Prepare search parameters
            params = {
                'originLocationCode': search_params['origin'],
                'destinationLocationCode': search_params['destination'],
                'departureDate': search_params['departure_date'],
                'adults': search_params['adults'],
                'travelClass': self._map_travel_class(search_params['travel_class'])
            }
            
            if search_params.get('return_date'):
                params['returnDate'] = search_params['return_date']
            
            # Search flights
            response = self.amadeus.shopping.flight_offers_search.get(**params)
            
            if not response.data:
                return []
            
            # Transform Amadeus response to our format
            flights = []
            for offer in response.data[:5]:  # Limit to 5 results
                flight_data = self._transform_flight_offer(offer, search_params)
                if flight_data:
                    flights.append(flight_data)
            
            return flights
            
        except ResponseError as error:
            logger.error(f"Amadeus API error: {error}")
            raise Exception(f"Flight search failed: {error}")
        except Exception as error:
            logger.error(f"Unexpected error in flight search: {error}")
            raise
    
    async def get_airport_info(self, airport_code: str) -> Dict[str, Any]:
        """Get airport information from Amadeus API."""
        try:
            response = self.amadeus.reference_data.locations.get(
                keyword=airport_code,
                subType='AIRPORT'
            )
            
            if not response.data:
                raise Exception(f"Airport not found: {airport_code}")
            
            airport = response.data[0]
            
            return {
                'code': airport.get('iataCode', airport_code),
                'name': airport.get('name', 'Unknown Airport'),
                'city': airport.get('address', {}).get('cityName', 'Unknown City'),
                'country': airport.get('address', {}).get('countryName', 'Unknown Country'),
                'timezone': airport.get('timeZoneOffset', 'Unknown'),
                'coordinates': {
                    'lat': float(airport.get('geoCode', {}).get('latitude', 0)),
                    'lon': float(airport.get('geoCode', {}).get('longitude', 0))
                }
            }
            
        except ResponseError as error:
            logger.error(f"Amadeus API error for airport {airport_code}: {error}")
            raise Exception(f"Airport info retrieval failed: {error}")
        except Exception as error:
            logger.error(f"Unexpected error getting airport info: {error}")
            raise
    
    def _map_travel_class(self, travel_class: str) -> str:
        """Map our travel class to Amadeus travel class."""
        mapping = {
            'economy': 'ECONOMY',
            'premium_economy': 'PREMIUM_ECONOMY',
            'business': 'BUSINESS',
            'first': 'FIRST'
        }
        return mapping.get(travel_class, 'ECONOMY')
    
    def _transform_flight_offer(self, offer: Any, search_params: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Transform Amadeus flight offer to our format."""
        try:
            if not offer.get('itineraries') or not offer.get('price'):
                return None
            
            itinerary = offer['itineraries'][0]
            segments = itinerary.get('segments', [])
            
            if not segments:
                return None
            
            first_segment = segments[0]
            last_segment = segments[-1]
            
            # Calculate total duration
            duration = itinerary.get('duration', 'PT0H0M')
            duration_formatted = self._format_duration(duration)
            
            # Get airline info
            airline_code = first_segment.get('carrierCode', 'XX')
            airline_name = self._get_airline_name(airline_code)
            
            # Get aircraft info
            aircraft = first_segment.get('aircraft', {}).get('code', 'Unknown')
            
            return {
                'airline': airline_name,
                'flight_number': f"{airline_code}{first_segment.get('number', '0000')}",
                'origin': search_params['origin'],
                'destination': search_params['destination'],
                'departure_date': search_params['departure_date'],
                'departure_time': self._format_time(first_segment.get('departure', {}).get('at', '')),
                'arrival_time': self._format_time(last_segment.get('arrival', {}).get('at', '')),
                'duration': duration_formatted,
                'price': {
                    'amount': float(offer['price']['total']),
                    'currency': offer['price']['currency'],
                    'per_person': float(offer['price']['total']) / search_params['adults']
                },
                'stops': len(segments) - 1,
                'travel_class': search_params['travel_class'],
                'aircraft': aircraft,
                'booking_class': first_segment.get('class', 'Y')
            }
            
        except Exception as error:
            logger.warning(f"Error transforming flight offer: {error}")
            return None
    
    def _format_duration(self, duration_str: str) -> str:
        """Format ISO 8601 duration to readable format."""
        try:
            # Parse PT2H30M format
            duration_str = duration_str.replace('PT', '')
            hours = 0
            minutes = 0
            
            if 'H' in duration_str:
                hours_part, remainder = duration_str.split('H')
                hours = int(hours_part)
                duration_str = remainder
            
            if 'M' in duration_str:
                minutes_part = duration_str.replace('M', '')
                if minutes_part:
                    minutes = int(minutes_part)
            
            return f"{hours}h {minutes}m"
        except:
            return "Unknown duration"
    
    def _format_time(self, datetime_str: str) -> str:
        """Extract time from datetime string."""
        try:
            if 'T' in datetime_str:
                return datetime_str.split('T')[1][:5]  # Get HH:MM
            return datetime_str[:5] if len(datetime_str) >= 5 else "00:00"
        except:
            return "00:00"
    
    def _get_airline_name(self, airline_code: str) -> str:
        """Get airline name from code."""
        airlines = {
            'AA': 'American Airlines',
            'DL': 'Delta Air Lines',
            'UA': 'United Airlines',
            'WN': 'Southwest Airlines',
            'B6': 'JetBlue Airways',
            'LH': 'Lufthansa',
            'BA': 'British Airways',
            'AF': 'Air France',
            'KL': 'KLM',
            'LX': 'Swiss International Air Lines'
        }
        return airlines.get(airline_code, f"Airline {airline_code}")